### PR TITLE
Test trips model

### PR DESCRIPTION
**Pull request by Showcas**
100% codecoverage for:
- trip model,
- ride model,
- drive model

Removed unused methods in ride

### CODE REVIEW

File: test/models/drive_test.dart
Comment (fynsta) [inline test/models/drive_test.dart:30]: I don't think that's a useful test, how could those types be wrong?

File: test/models/drive_test.dart
Comment (fynsta) [inline test/models/drive_test.dart:791]: Do we want to write these tests for every model... I feel like that's just a few hundred lines of code that is not particularly insightful

File: test/models/drive_test.dart
Comment (fynsta) [inline test/models/drive_test.dart:200]: can you check the language here and everywhere else? this is just not proper grammar and I'm not even sure what it's trying to say

**Comment by fynsta** [comment]
can you please stop using the normal "git pull" to merge the branches (or whatever you are doing, I'm not completely sure...). The commit history looks pretty broken, and I can't rebase this branch on main now.

**Comment by Showcas** [comment]
> can you please stop using the normal "git pull" to merge the branches (or whatever you are doing, I'm not completely sure...). The commit history looks pretty broken, and I can't rebase this branch on main now.

I used the normal pull and rebase

File: test/models/drive_test.dart
Comment (Showcas) [reply test/models/drive_test.dart:200]: ahh yes sorry

File: test/models/drive_test.dart
Comment (Showcas) [reply test/models/drive_test.dart:30]: ok

**Comment by fynsta** [comment]
> I used the normal pull to get main and then I use rebase. how else am I suppose to do it? merge hab ich seit deiner letzten bitte nicht mehr benutzt

Oh okay then I'm sorry. I'm confused why the git rebase behaviour is broken in that case... And this commit looks suspiciously like a merge commit: https://github.com/MOTIS-Mitfahr-App/flutter-app/pull/80/commits/828345150f6dc6f4cc449f13187fce1de9c96f6e

**Comment by fynsta** [comment]
And I think it's also why Codecov does not work for this PR. It cannot find a base commit. Not sure what happened

**Comment by Showcas** [comment]
> > I used the normal pull to get main and then I use rebase. how else am I suppose to do it? merge hab ich seit deiner letzten bitte nicht mehr benutzt
> 
> Oh okay then I'm sorry. I'm confused why the git rebase behaviour is broken in that case... And this commit looks suspiciously like a merge commit: [8283451](https://github.com/MOTIS-Mitfahr-App/flutter-app/commit/828345150f6dc6f4cc449f13187fce1de9c96f6e)

I noticed it aswell mhhh. We can talk later about it maybe Im doing something wrong

File: test/models/ride_test.dart
Comment (istzen) [inline test/models/ride_test.dart:269]: do we really need to test these methods for ride and drive?

Comment (fynsta) [reply test/models/ride_test.dart:269]: I also think that they should be moved to some trip_test.
 
But as a note: There you can only use a model of a specific type... So if I tested only for ride in that file, then I would not catch it if the behaviour were to be changed for drives in the future.

File: test/models/drive_test.dart
Comment (Showcas) [reply test/models/drive_test.dart:750]: Should I reduce the tests for equal? Or should we just leave them there since they already exist

Comment (fynsta) [reply test/models/drive_test.dart:750]: Let's leave them here, they don't hurt right now

File: test/models/trip_test.dart
Comment (fynsta) [inline test/models/trip_test.dart:18]: okay but now can you write "trip" everywhere instead of ride/drive

Comment (Showcas) [reply test/models/trip_test.dart:18]: I renamed the variables. But I would leave the description cause they show that we cover rides and drives

Comment (fynsta) [reply test/models/trip_test.dart:18]: I meant in the group and test description

File: test/models/trip_test.dart
Comment (fynsta) [inline test/models/trip_test.dart:18]: "handel" -> "handle"
"finsihed" -> "finished" (down below somewhere)

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:17]: Please remove these comments

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:148]:
```suggestion
  group('Drive.approvedRides', () {
```

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:163]: Do we need this test? One approved ride is already the test below this, and I don't have reason to expect the outcome of this to be different

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:183]: Please keep the test cases minimal. They should document the code and be readable. Every test case should be specific. Here, for example, there is no need for one Ride of every other status to be in the list, since it that has already been tested in the test above.

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:233]: > Please keep the test cases minimal. They should document the code and be readable. Every test case should be specific. Here, for example, there is no need for one Ride of every other status to be in the list, since it that has already been tested in the test above.

Here, too

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:271]: This list can be shortened

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:331]: Please remove this almost exact copy

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:366]: IMO, all of these further rides are unnecessary. We have established in the previous test that every ride with a status != approved will be counted as having 0 seats, so this is clutter

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:396]: This, too. You want to establish that for rides at the same time, seats will be summed up. So I'd just create two rides at the same time, maybe one at a different time for good measure (even though that would arguably be a different test). Remove the rest, please. As a maintainer, this would be very time-consuming to read in order to understand the relatively basic message scenario this test conveys/tests

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:429]: I think in this case the point can be made that it's fair to test a more complex case, but please at least make a comment which rides overlap here so that I don't have to read all of this code (DateTimes aren't that easy to read)

Comment (Eddie-42) [reply test/models/drive_test.dart:429]: Also, and I think this goes for all of these tests, the description isn't really ideal, IMO. I would think about what I'd want to see in a console if a test fails to know exactly where it has failed, quickly. In this case, I'm not interested how many seats there are and what is expected; this information is already provided in the code once I go look for the test. What I'd want to know is that the complex Drive.getMaxUsedSeats test has failed. In the test above, I'd want to know that the overlapping getMaxUsedSeats test has failed, and in the test above that I'd want to know that the non-overlapping getMaxUsedSeats test has failed. I think naming tests like this might already help determining useful test cases

**Review by Eddie-42 — CHANGES_REQUESTED** [review]
I think I've made my point regarding the tests using some examples. The same points could be made about some other tests, but I don't want to repeat myself too often. Please revise the tests from the point of view that it should be documenting the code and just as readable as the code. If anyone wants to discuss that with me, I'm willing to defend my position :) But 2000 lines is definitely too much for these tests IMO

File: test/models/drive_test.dart
Comment (Showcas) [reply test/models/drive_test.dart:63]: I forgot about that one

**Comment by Showcas** [comment]
> I think I've made my point regarding the tests using some examples. The same points could be made about some other tests, but I don't want to repeat myself too often. Please revise the tests from the point of view that it should be documenting the code and just as readable as the code. If anyone wants to discuss that with me, I'm willing to defend my position :) But 2000 lines is definitely too much for these tests IMO

I have redone the tests and tried to to take your points into account

File: test/models/trip_test.dart
Comment (fynsta) [inline test/models/trip_test.dart:22]: the return type could be `TripFactory`, then you would not need the casting every time

File: test/models/trip_test.dart
Comment (fynsta) [inline test/models/trip_test.dart:17]: I would remove this comment everywhere, it's not even true

File: test/util/factories/trip_factory.dart
Comment (fynsta) [inline test/util/factories/trip_factory.dart:19]: can you move this to the spot where it is in the class definition? Definitely above `createDependencies`, that should always be the last argument for clarity

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:40]:
```suggestion
    test('trip is finished', () {
```

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:204]: Am I wrong or does this not have a test case that is actually equal? :D

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:174]: This kind of sounds like a widget test. I would try to go for something more descriptive of the test case, like "future trip in future list view" or something

Comment (Eddie-42) [reply test/models/trip_test.dart:174]: Also, what is the difference between this and the `Trip.shouldShowInListView` group before?

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:144]:
```suggestion
    test('returns true for past finished trip', () {
```

Please use a code spell checker :) I'm not going to comment on spelling mistakes anymore from here on out

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:94]: What's with trip is before range?

File: test/models/trip_test.dart
Comment (Eddie-42) [inline test/models/trip_test.dart:74]: What's with trip is after other?

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:151]: IMO, what you could very well do here is have this list contain rides from every status EXCEPT approved, to prove the point that only approved rides will be counted. I think that could even be a neat one-liner, like 
```
rides: List.generate(
            RideStatus.values.length, (index) => RideFactory().generateFake(status: RideStatus.values[index]))
            .where((element) => element.status != RideStatus.approved)
            .toList(),
```

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:162]: Same here and in `Drive.pendingRides`

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:276]: Could you please continue using `DateTime.now()` here and everywhere else? I think this is really illegible

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:340]: Please remove `can handle` from these descriptions. You could write these before any test case and I don't think it adds any value (and in this case, it's inconsistent with the description right below)

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:448]: I think you should use a deepcopy here. That would show that we're using value equality, not reference equality. For Ride too

File: test/models/drive_test.dart
Comment (Eddie-42) [inline test/models/drive_test.dart:454]: I don't like this test. Either we test this through for every parameter (which I wouldn't do), or we just do it for one, and I think that's preferably the ID. The case can be made that it is important to show that even drives with the same ID are different to us if they have different values, but then why cancelled? I think start or something less obscure would make more sense in that case (you already have driverId below, that's fine too)

File: test/models/ride_test.dart
Comment (Eddie-42) [inline test/models/ride_test.dart:178]: Do we test this now? I think tests for testing infrastructure are a bit much, we're already getting the coverage by calling this 100 times in tests anyway.

File: test/models/ride_test.dart
Comment (Eddie-42) [inline test/models/ride_test.dart:287]:
```suggestion
    test('show in ListView', () {
```
Also, isn't this already covered by the trip test?

File: test/models/ride_test.dart
Comment (Eddie-42) [inline test/models/ride_test.dart:333]: Here you're testing this for a few random parameters again. I would suggest testing it for the ID and maybe one additional parameter, (or, if you want to test every parameter, do that and do it in a loop, please)

**Review by Eddie-42 — COMMENTED** [review]
I think I am very harsh in this review, but I do think this is a great improvement over the previous iteration and after this is all fixed, I am happy to approve. Very good job!

File: test/models/trip_test.dart
Comment (Showcas) [reply test/models/trip_test.dart:22]: It does not see RideFactory or RideFactory as an instance of TripFactory when I declare TripFactory as a Return type. The reason could be since they don't have a constructor.

File: test/models/trip_test.dart
Comment (Showcas) [reply test/models/trip_test.dart:17]: missed that one, removed it in all the others

File: test/models/trip_test.dart
Comment (Showcas) [reply test/models/trip_test.dart:204]: Forgot to add that one here

File: test/models/trip_test.dart
Comment (Showcas) [reply test/models/trip_test.dart:174]: forgot to remove the old one

File: test/models/drive_test.dart
Comment (Showcas) [reply test/models/drive_test.dart:454]: that is covered by the equals test in trip. since euquals in drive and ride call their super from trip. I only check for the extra checks in drive and ride and the core trip values in Trip.equals

File: test/models/ride_test.dart
Comment (Showcas) [reply test/models/ride_test.dart:178]: I can take it out if you want. But I don't think it hurts

File: test/models/ride_test.dart
Comment (Showcas) [reply test/models/ride_test.dart:287]: here are new cases since it is overwriten in ride. In drive it is not overwriten so it is not tested in drive

File: test/util/factories/mock_trip_factory.dart
Comment (fynsta) [inline test/util/factories/mock_trip_factory.dart:6]: I'm not feeling great about two things: 

- this file/factory does not fit with the others
- the fact that we need to repeat all fields here _again_... 

Maybe the best solution is just to remove the `abstract` keyword from Trip, and then use the `TripFactory` for our purposes. I think that would solve a lot of complication, and `abstract` is currently really unnecessary since `Trip` implements all its methods.

What do you think @Eddie-42 ?

Comment (fynsta) [reply test/util/factories/mock_trip_factory.dart:6]: I have pushed my idea on another branch. Please see this commit: https://github.com/MOTIS-Mitfahr-App/flutter-app/commit/283025be1f69fc9670e73eb2b3ae855b10b15adc

I think it's a lot better?

Comment (Showcas) [reply test/util/factories/mock_trip_factory.dart:6]: I agree

File: test/models/trip_test.dart
Comment (istzen) [inline test/models/trip_test.dart:11]: can you rebase? then this needs to be a MockRequestProcessor if i am correct

Comment (fynsta) [reply test/models/trip_test.dart:11]: as written above the rebase is broken 😢  Maybe I'll have to open a new branch and copy the changes by hand.

File: test/models/drive_test.dart
Comment (Eddie-42) [reply test/models/drive_test.dart]: Mh okay I suspect we'll have to unify that anyway some time

File: test/models/drive_test.dart
Comment (Eddie-42) [reply test/models/drive_test.dart:270]: Think you forgot that one because GitHub kind of hides it ;)

**Review by Eddie-42 — APPROVED** [review]

File: test/models/trip_test.dart
Comment (istzen) [reply test/models/trip_test.dart:11]: ah sorry my mistake. You could do that i guess or is there any other way?

Comment (istzen) [reply test/models/trip_test.dart:11]: maybe that's the best way idk if there is another one

**Review by fynsta — APPROVED** [review]

**Review by MarkFalke — APPROVED** [review]

**Comment by fynsta** [comment]
Superseded by #97

**Pull request closed without merging**
